### PR TITLE
IOI Sails Solution

### PR DESCRIPTION
**Pull request by eysbutno**
_Place an "x" in the corresponding checkbox if it is done or does not apply to this pull request._

- [x] I have tested my code.
- [x] I have added my solution according to the steps [here](https://usaco.guide/general/adding-solution#steps).
- [x] I have followed the code conventions mentioned [here](https://usaco.guide/general/adding-solution/#code-conventions).
  - I understand that if it is clear that I have not attempted to follow these conventions, my PR will be closed.
  - If changes are requested, I will re-request a review after addressing them.
- [x] I have linked this PR to any issues that it closes.

will probably make the explanation clearer tmrw, as my brain is thoroughly cooked rn

### CODE REVIEW

File: solutions/platinum/ioi-07-sails.mdx [hunk 1]
```diff
@@ -0,0 +1,138 @@
...
+To remedy this issue, we split up our range addition into two separate updates.
+In the case above, we can just split our previous range addition into updates
+on the range $[1, 1]$ and $[3, 5]$.
+
+## Implementation
+
+**Time Complexity:** $\mathcal{O}(N\log^2N)$
+
+Note: It is possible to cut a log factor by walking on the BIT, as described in
+[this](https://codeforces.com/blog/entry/61364) blog.
```
Comment (SansPapyrus683) [inline solutions/platinum/ioi-07-sails.mdx:56]: you can put this in an info tag

File: solutions/platinum/ioi-07-sails.mdx [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+	T pref_sum(int ind) {
+		ind++;
+		T total = 0;
+		for (; ind > 0; ind -= ind & -ind) { total += bit[ind]; }
+		return total;
+	}
+};
+// EndCodeSnip
+
+template <typename T, typename F> T first_true(T low, T high, const F &fn) {
```
Comment (SansPapyrus683) [inline solutions/platinum/ioi-07-sails.mdx:93]: why template this time

File: solutions/platinum/ioi-07-sails.mdx [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+	int n;
+	cin >> n;
+	vector<int> h(n), k(n);
+	for (int i = 0; i < n; i++) { cin >> h[i] >> k[i]; }
+
+	vector<int> ord(n);
+	iota(begin(ord), end(ord), 0);
+	sort(begin(ord), end(ord), [&](int i, int j) -> bool { return h[i] < h[j]; });
+
+	const int MX = *max_element(begin(h), end(h));
```
Comment (SansPapyrus683) [inline solutions/platinum/ioi-07-sails.mdx:111]: well this isn't really an *immutable* constant, so i'd just put it as max_x or whatever

**Review by SansPapyrus683 — CHANGES_REQUESTED** [review]
peak

File: solutions/platinum/ioi-07-sails.mdx [hunk 4]
```diff
@@ -0,0 +1,138 @@
...
+a binary indexed tree maintaining the number of sails at each level in descending
+order. We place all sails on the range $[h[i] - k[i] + 1, h[i]]$ (in the BIT),
+as that is optimal. To handle placing a sail on every index in the range, we
+use the difference array idea on our BIT.
+
+However, applying this range addition may cause the values in the BIT to be
+sorted in descending order. Let's assume that after applying all the updates,
+the values in the BIT are the following:
+
+$[5, 5, 4, 3, 1]$
```
Comment (SansPapyrus683) [inline solutions/platinum/ioi-07-sails.mdx:40]: also i think double dollar signs would be more appropriate here

File: solutions/platinum/ioi-07-sails.mdx (hunk 2 above)
Comment (TheGamingMousse) [reply solutions/platinum/ioi-07-sails.mdx:95]: imo it makes the code cleaner & makes it shorter

File: solutions/platinum/ioi-07-sails.mdx [hunk 5]
```diff
@@ -0,0 +1,140 @@
+---
+id: ioi-07-sails
+source: IOI
+title: 2007 - Sails
+author: Justin Ji
+---
+
+[Official Analysis](https://dmoj.ca/problem/ioi07p3/editorial)
```
Comment (ryanchou-dev) [inline solutions/platinum/ioi-07-sails.mdx:8]: cooked

File: solutions/platinum/ioi-07-sails.mdx (hunk 2 above)
Comment (ryanchou-dev) [reply solutions/platinum/ioi-07-sails.mdx:95]: does this really need to be a template though 😭 

also typename t not needed

**Review by ryanchou-dev — COMMENTED** [review]
🤩

File: solutions/platinum/ioi-07-sails.mdx (hunk 2 above)
Comment (TheGamingMousse) [reply solutions/platinum/ioi-07-sails.mdx:95]: fair enough, i'll write a function inside main to do the same thing in a sec

File: solutions/platinum/ioi-07-sails.mdx (hunk 5 above)
Comment (TheGamingMousse) [reply solutions/platinum/ioi-07-sails.mdx:8]: i FINALLY REMEMBERED

File: solutions/platinum/ioi-07-sails.mdx [hunk 6]
```diff
@@ -0,0 +1,142 @@
...
+		int idx_2 = first_val(val + 1, h[i]);
+		bit.add(idx_1, 1);
+		bit.add(h[i], -1);
+		bit.add(idx_2, 1);
+		bit.add(idx_2 + k[i] - (h[i] - idx_1), -1);
+	}
+
+	ll res = 0;
+	for (int i = 0; i < max_h; i++) {
+		int cnt = bit.pref_sum(i);
```
Comment (SansPapyrus683) [inline solutions/platinum/ioi-07-sails.mdx:132]: use another name

Comment (TheGamingMousse) [reply solutions/platinum/ioi-07-sails.mdx]: changed it to `sail_cnt`, is that better?

**Review by SansPapyrus683 — APPROVED** [review]

**Review by ryanchou-dev — APPROVED** [review]

**Comment by ryanchou-dev** [comment]
good work!

**Pull request merged**